File: chapter-06-metadata/java/01-termstore/TermStoreExplorer.java
// TermStoreExplorer.java
// Ejemplo: Explorar el Term Store con Microsoft Graph SDK para Java (v6).
// El GraphServiceClient se inyecta por constructor (DI) desde el módulo común.

package com.sharepointexperiencia.managedmetadata;

import com.laexperiencia.sharepoint.common.GraphServiceClientFactory;
import com.microsoft.graph.serviceclient.GraphServiceClient;

public class TermStoreExplorer {
    private final GraphServiceClient graphClient;

    public TermStoreExplorer(GraphServiceClient graphClient) {
        this.graphClient = graphClient;
    }

    public void explore() {
        System.out.println("=".repeat(60));
        System.out.println("Explorando Term Store");
        System.out.println("=".repeat(60));
        System.out.println();

        try {
            // Obtener Term Store (API v6: sites().bySiteId(...).termStore())
            var termStore = graphClient.sites().bySiteId("root").termStore().get();

            System.out.println("📚 Term Store Info:");
            System.out.println("   ID: " + termStore.getId());
            System.out.println("   Default Language: " + termStore.getDefaultLanguageTag());
            System.out.println();

            // Listar grupos
            var groups = graphClient.sites().bySiteId("root").termStore().groups().get();

            int groupCount = groups != null && groups.getValue() != null ? groups.getValue().size() : 0;
            System.out.println("📂 Grupos encontrados: " + groupCount + "\n");

            if (groups != null && groups.getValue() != null) {
                for (var group : groups.getValue()) {
                    System.out.println("   📁 " + group.getDisplayName());
                    System.out.println("      ID: " + group.getId());
                    String desc = group.getDescription() != null ? group.getDescription() : "N/A";
                    System.out.println("      Descripción: " + desc);

                    // Obtener conjuntos de términos
                    var sets = graphClient.sites().bySiteId("root")
                            .termStore().groups().byGroupId(group.getId()).sets().get();

                    if (sets != null && sets.getValue() != null) {
                        for (var set : sets.getValue()) {
                            String setName = (set.getLocalizedNames() != null && !set.getLocalizedNames().isEmpty())
                                    ? set.getLocalizedNames().get(0).getName()
                                    : set.getId();
                            System.out.println("         📚 " + setName);
                        }
                    }
                    System.out.println();
                }
            }
        } catch (Exception e) {
            System.err.println("❌ Error: " + e.getMessage());
        }
    }

    public static void main(String[] args) {
        try {
            // create() auto-detecta certificado (si CERTIFICATE_PATH está presente) o client secret.
            var client = GraphServiceClientFactory.create();
            new TermStoreExplorer(client).explore();
            System.out.println("\nTerm Store exploration completed.");
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            System.exit(1);
        }
    }
}

File: chapter-06-metadata/csharp/01-termstore/TermStoreExplorer.cs
// TermStoreExplorer.cs
// Example: Explore the SharePoint Term Store using Microsoft Graph SDK.
// The Graph client is injected via constructor (DI) from the common module.

using LaExperiencia.SharePoint.Common;
using Microsoft.Graph;
using Microsoft.Graph.Models;

namespace LaExperiencia.SharePoint.Chapter06.Metadata;

class TermStoreExplorer
{
    private readonly GraphServiceClient _graphClient;

    public TermStoreExplorer(GraphServiceClient graphClient)
    {
        _graphClient = graphClient;
    }

    public async Task ExploreAsync()
    {
        Console.WriteLine("=" .Repeat(60));
        Console.WriteLine("Explorando Term Store");
        Console.WriteLine("=".Repeat(60));
        Console.WriteLine();

        try
        {
            // Get Term Store information
            var termStore = await _graphClient.Sites["root"]
                .TermStore
                .GetAsync();

            Console.WriteLine("📚 Term Store Info:");
            Console.WriteLine($"   ID: {termStore?.Id}");
            Console.WriteLine($"   Default Language: {termStore?.DefaultLanguageTag}");
            Console.WriteLine($"   Languages: {string.Join(", ", termStore?.LanguageTags ?? new List<string>())}");
            Console.WriteLine();

            // List groups
            var groups = await _graphClient.Sites["root"]
                .TermStore
                .Groups
                .GetAsync();

            Console.WriteLine($"📂 Grupos encontrados: {groups?.Value?.Count ?? 0}\n");

            foreach (var group in groups?.Value ?? new List<Microsoft.Graph.Models.TermStore.Group>())
            {
                Console.WriteLine($"   📁 {group.DisplayName}");
                Console.WriteLine($"      ID: {group.Id}");
                Console.WriteLine($"      Description: {group.Description ?? "N/A"}");

                // Get term sets
                var sets = await _graphClient.Sites["root"]
                    .TermStore
                    .Groups[group.Id]
                    .Sets
                    .GetAsync();

                foreach (var set in sets?.Value ?? new List<Microsoft.Graph.Models.TermStore.Set>())
                {
                    var setName = set.LocalizedNames?.FirstOrDefault()?.Name ?? set.Id;
                    Console.WriteLine($"         📚 {setName}");
                }
                Console.WriteLine();
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"❌ Error: {ex.Message}");
        }
    }

    /// <summary>
    /// Entry point: builds the client with the common module and explores the Term Store (read-only).
    /// </summary>
    public static async Task Main(string[] args)
    {
        try
        {
            var graphClient = SharePointGraphClientFactory.CreateFromSecret();
            var explorer = new TermStoreExplorer(graphClient);
            await explorer.ExploreAsync();
            Console.WriteLine("\nTerm Store exploration completed.");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error: {ex.Message}");
            Environment.Exit(1);
        }
    }
}

// Extension for string.Repeat
public static class StringExtensions
{
    public static string Repeat(this string s, int count)
    {
        return string.Concat(Enumerable.Repeat(s, count));
    }
}
